libs/dal/src/types/mongo.ts

```diff
@@ -0,0 +1,69 @@
+import { Condition } from 'mongoose';
+
+import { LeafKeys } from './helpers';
+
+/**
+ * Modified from original Mongoose typings to introduce stricter typing
+ * for the `FilterQuery` generic type
+ *
+ * @see https://github.com/Automattic/mongoose/blob/582156858db3ca7fbaa8950dc997e0d9e8117b21/types/query.d.ts#L70
+ */
+type RootQuerySelector<T> = {
+  /** @see https://www.mongodb.com/docs/manual/reference/operator/query/and/#op._S_and */
+  $and?: Array<FilterQuery<T>>;
+  /** @see https://www.mongodb.com/docs/manual/reference/operator/query/nor/#op._S_nor */
+  $nor?: Array<FilterQuery<T>>;
+  /** @see https://www.mongodb.com/docs/manual/reference/operator/query/or/#op._S_or */
+  $or?: Array<FilterQuery<T>>;
+  /** @see https://www.mongodb.com/docs/manual/reference/operator/query/text */
+  $text?: {
+    $search: string;
+    $language?: string;
+    $caseSensitive?: boolean;
+    $diacriticSensitive?: boolean;
+  };
+  /** @see https://www.mongodb.com/docs/manual/reference/operator/query/where/#op._S_where */
+  $where?: string;
+  /** @see https://www.mongodb.com/docs/manual/reference/operator/query/comment/#op._S_comment */
+  $comment?: string;
+} & Partial<{
+  [Key in LeafKeys<T>]: T[Key] | T[Key][];
```

Comment:
`Key` can be a dot notation of a leaf field - this still needs some work to make sure the `T[Key]` resolves to the correct type.